### PR TITLE
Add the share download hook to files_sharing

### PR DESCRIPTION
**Pull request by brnskn**
In some cases, managing the share downloading is needed but, there is no any different hook for it. That's why I added a "share_link_download" hook to ShareController. 

Related issues: #19546

Signed-off-by: Baran Sekin <baransekin@gmail.com>

### CODE REVIEW

File: apps/files_sharing/lib/Controller/ShareController.php [hunk 1]
```diff
@@ -593,6 +622,7 @@ public function downloadShare($token, $files = null, $path = '', $downloadStartS
 					$node = $node->get($path);
 				} catch (NotFoundException $e) {
 					$this->emitAccessShareHook($share, 404, 'Share not found');
+					$this->emitDownloadShareHook($share, 404, 'Share not found');
```
Comment (kesselb) [inline apps/files_sharing/lib/Controller/ShareController.php:625]: Mind to explain? File is not found so there is no download ;)

**Review by kesselb — COMMENTED** [review]
- Mind to add a use case to your description? Just to have some documentation what it's good for. 
- We are migrating away from hooks to a event based system. Not sure if adding a new hook is possible.

File: apps/files_sharing/lib/Controller/ShareController.php (hunk 1 above)
Comment (brnskn) [reply apps/files_sharing/lib/Controller/ShareController.php:625]: I've followed the other accessShare hook standards. Maybe this information can be useful like "accessShare share not found" hook

**Comment by brnskn** [comment]
> * Mind to add a use case to your description? Just to have some documentation what it's good for.
> * We are migrating away from hooks to a event based system. Not sure if adding a new hook is possible.

For the use case description:

I have developed an app that uses this download hook and I can limit the share download count for security reasons. When the share is downloaded once, I'm deleting that share automatically using ShareManager.

**Comment by szaimen** [comment]
> I have developed an app that uses this download hook and I can limit the share download count for security reasons. When the share is downloaded once, I'm deleting that share automatically using ShareManager. And also that feature is disabled by default but can be enabled from the share link settings.

Sounds great, imo!
Does it also work for videos? (I know of some reports, where it was reported, that watching a video reports multiple download activity entries which could be a problem here)
BTW: if that is actually providing one time download links, it would resolve https://github.com/nextcloud/server/issues/6841

**Comment by brnskn** [comment]
> > I have developed an app that uses this download hook and I can limit the share download count for security reasons. When the share is downloaded once, I'm deleting that share automatically using ShareManager. And also that feature is disabled by default but can be enabled from the share link settings.
> 
> Sounds great, imo!
> Does it also work for videos? (I know of some reports, where it was reported, that watching a video over a public link reports multiple download activity entries which could be a problem here)
> BTW: if that is actually providing one time download links, it would resolve #6841

Yes, if this pr is merged, I will make open source that one-time download application :)

**Comment by blizzz** [comment]
we're going away from OC_Hook, better use the eventdispatcher right away. I am not sure this is sufficient for webdav.

**Comment by skjnldsv** [comment]
> we're going away from OC_Hook, better use the eventdispatcher right away.

Closing

**Comment by skjnldsv** [comment]
We're also in the process of implementing one-time download for 29 :)

**Pull request closed without merging**